File: backend/routes/requests.js
const express = require('express');
const router = express.Router();
const pool = require('../config/database');

// GET all pending substitute requests
router.get('/', async (req, res) => {
  try {
    const result = await pool.query(
      `SELECT sr.*, u.name as teacher_name 
       FROM substitute_requests sr 
       JOIN users u ON sr.teacher_id = u.id 
       WHERE sr.status = 'pending' 
       ORDER BY sr.date ASC, sr.time ASC`
    );
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching requests:', error);
    res.status(500).json({ error: 'Failed to fetch requests' });
  }
});

// GET requests by teacher
router.get('/teacher/:teacherId', async (req, res) => {
  try {
    const { teacherId } = req.params;
    const result = await pool.query(
      `SELECT sr.*, u.name as acceptor_name 
       FROM substitute_requests sr 
       LEFT JOIN users u ON sr.accepted_by = u.id 
       WHERE sr.teacher_id = $1 
       ORDER BY sr.created_at DESC`,
      [teacherId]
    );
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching teacher requests:', error);
    res.status(500).json({ error: 'Failed to fetch teacher requests' });
  }
});

// POST create new substitute request
router.post('/', async (req, res) => {
  try {
    const { teacher_id, subject, date, time, duration, classroom, notes } = req.body;

    // Validate required fields
    if (!teacher_id || !subject || !date || !time || !duration || !classroom) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    const result = await pool.query(
      `INSERT INTO substitute_requests 
       (teacher_id, subject, date, time, duration, classroom, notes) 
       VALUES ($1, $2, $3, $4, $5, $6, $7) 
       RETURNING *`,
      [teacher_id, subject, date, time, duration, classroom, notes]
    );

    res.status(201).json(result.rows[0]);
  } catch (error) {
    console.error('Error creating request:', error);
    res.status(500).json({ error: 'Failed to create request' });
  }
});

// PUT accept a substitute request
router.put('/:id/accept', async (req, res) => {
  try {
    const { id } = req.params;
    const { teacher_id } = req.body;

    if (!teacher_id) {
      return res.status(400).json({ error: 'teacher_id is required' });
    }

    // Check if request exists and is pending
    const checkResult = await pool.query(
      'SELECT * FROM substitute_requests WHERE id = $1',
      [id]
    );

    if (checkResult.rows.length === 0) {
      return res.status(404).json({ error: 'Request not found' });
    }

    if (checkResult.rows[0].status !== 'pending') {
      return res.status(400).json({ error: 'Request is not available' });
    }

    // Accept the request
    const result = await pool.query(
      `UPDATE substitute_requests 
       SET status = 'accepted', accepted_by = $1, updated_at = CURRENT_TIMESTAMP 
       WHERE id = $2 
       RETURNING *`,
      [teacher_id, id]
    );

    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error accepting request:', error);
    res.status(500).json({ error: 'Failed to accept request' });
  }
});

// PUT cancel a substitute request
router.put('/:id/cancel', async (req, res) => {
  try {
    const { id } = req.params;
    const { teacher_id } = req.body;

    // Check if the request belongs to the teacher
    const checkResult = await pool.query(
      'SELECT * FROM substitute_requests WHERE id = $1 AND teacher_id = $2',
      [id, teacher_id]
    );

    if (checkResult.rows.length === 0) {
      return res.status(404).json({ error: 'Request not found or unauthorized' });
    }

    const result = await pool.query(
      `UPDATE substitute_requests 
       SET status = 'cancelled', updated_at = CURRENT_TIMESTAMP 
       WHERE id = $1 
       RETURNING *`,
      [id]
    );

    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error cancelling request:', error);
    res.status(500).json({ error: 'Failed to cancel request' });
  }
});

// DELETE a substitute request
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { teacher_id } = req.body;

    // Check if the request belongs to the teacher
    const checkResult = await pool.query(
      'SELECT * FROM substitute_requests WHERE id = $1 AND teacher_id = $2',
      [id, teacher_id]
    );

    if (checkResult.rows.length === 0) {
      return res.status(404).json({ error: 'Request not found or unauthorized' });
    }

    await pool.query('DELETE FROM substitute_requests WHERE id = $1', [id]);
    res.json({ message: 'Request deleted successfully' });
  } catch (error) {
    console.error('Error deleting request:', error);
    res.status(500).json({ error: 'Failed to delete request' });
  }
});

module.exports = router;

File: backend/routes/requests.py
from fastapi import APIRouter, HTTPException, status
from typing import List

from database import get_supabase
from models import (
    SubstituteRequestCreate,
    SubstituteRequestResponse,
    SubstituteRequestUpdate,
    AcceptRequest,
    CancelRequest
)
from services.push_notifications import notify_all_faculty_except, notify_user

router = APIRouter()


@router.get("/", response_model=List[SubstituteRequestResponse])
async def get_pending_requests():
    """
    Get all pending substitute requests.
    Returns requests ordered by date and time.
    """
    supabase = get_supabase()
    
    try:
        # Get pending requests with teacher name
        result = supabase.table("substitute_requests")\
            .select("*, users!substitute_requests_teacher_id_fkey(name)")\
            .eq("status", "pending")\
            .order("date")\
            .order("time")\
            .execute()
        
        requests_list = []
        for req in result.data:
            teacher_name = None
            if req.get("users"):
                teacher_name = req["users"].get("name")
            
            requests_list.append(SubstituteRequestResponse(
                id=req["id"],
                teacher_id=req["teacher_id"],
                subject=req["subject"],
                date=req["date"],
                time=req["time"],
                duration=req["duration"],
                classroom=req["classroom"],
                notes=req.get("notes"),
                status=req["status"],
                accepted_by=req.get("accepted_by"),
                created_at=req.get("created_at"),
                updated_at=req.get("updated_at"),
                teacher_name=teacher_name
            ))
        
        return requests_list
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch requests: {str(e)}"
        )


@router.get("/teacher/{teacher_id}", response_model=List[SubstituteRequestResponse])
async def get_teacher_requests(teacher_id: int):
    """
    Get all substitute requests created by a specific teacher.
    """
    supabase = get_supabase()
    
    try:
        # Get requests with acceptor name
        result = supabase.table("substitute_requests")\
            .select("*, acceptor:users!substitute_requests_accepted_by_fkey(name)")\
            .eq("teacher_id", teacher_id)\
            .order("created_at", desc=True)\
            .execute()
        
        requests_list = []
        for req in result.data:
            acceptor_name = None
            if req.get("acceptor"):
                acceptor_name = req["acceptor"].get("name")
            
            requests_list.append(SubstituteRequestResponse(
                id=req["id"],
                teacher_id=req["teacher_id"],
                subject=req["subject"],
                date=req["date"],
                time=req["time"],
                duration=req["duration"],
                classroom=req["classroom"],
                notes=req.get("notes"),
                status=req["status"],
                accepted_by=req.get("accepted_by"),
                created_at=req.get("created_at"),
                updated_at=req.get("updated_at"),
                acceptor_name=acceptor_name
            ))
        
        return requests_list
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch teacher requests: {str(e)}"
        )


@router.get("/{request_id}", response_model=SubstituteRequestResponse)
async def get_request(request_id: int):
    """
    Get a specific substitute request by ID.
    """
    supabase = get_supabase()
    
    try:
        result = supabase.table("substitute_requests")\
            .select("*, teacher:users!substitute_requests_teacher_id_fkey(name), acceptor:users!substitute_requests_accepted_by_fkey(name)")\
            .eq("id", request_id)\
            .execute()
        
        if not result.data or len(result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        req = result.data[0]
        teacher_name = req["teacher"]["name"] if req.get("teacher") else None
        acceptor_name = req["acceptor"]["name"] if req.get("acceptor") else None
        
        return SubstituteRequestResponse(
            id=req["id"],
            teacher_id=req["teacher_id"],
            subject=req["subject"],
            date=req["date"],
            time=req["time"],
            duration=req["duration"],
            classroom=req["classroom"],
            notes=req.get("notes"),
            status=req["status"],
            accepted_by=req.get("accepted_by"),
            created_at=req.get("created_at"),
            updated_at=req.get("updated_at"),
            teacher_name=teacher_name,
            acceptor_name=acceptor_name
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch request: {str(e)}"
        )


@router.post("/", response_model=SubstituteRequestResponse, status_code=status.HTTP_201_CREATED)
async def create_request(request: SubstituteRequestCreate):
    """
    Create a new substitute request.
    """
    supabase = get_supabase()
    
    try:
        # Verify teacher exists
        teacher_result = supabase.table("users").select("id, name").eq("id", request.teacher_id).execute()
        
        if not teacher_result.data or len(teacher_result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Teacher not found"
            )
        
        teacher_name = teacher_result.data[0]["name"]
        
        # Create request
        new_request = {
            "teacher_id": request.teacher_id,
            "subject": request.subject,
            "date": str(request.date),
            "time": request.time,
            "duration": request.duration,
            "classroom": request.classroom,
            "notes": request.notes,
            "status": "pending"
        }
        
        result = supabase.table("substitute_requests").insert(new_request).execute()
        
        if not result.data or len(result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create request"
            )
        
        req = result.data[0]
        
        # Send push notification to all other faculty
        await notify_all_faculty_except(
            exclude_user_id=request.teacher_id,
            title="📚 New Substitute Request",
            body=f"{teacher_name} needs a substitute for {request.subject} on {request.date} at {request.time}",
            data={
                "type": "new_request",
                "request_id": req["id"],
                "subject": request.subject,
            }
        )
        
        return SubstituteRequestResponse(
            id=req["id"],
            teacher_id=req["teacher_id"],
            subject=req["subject"],
            date=req["date"],
            time=req["time"],
            duration=req["duration"],
            classroom=req["classroom"],
            notes=req.get("notes"),
            status=req["status"],
            accepted_by=req.get("accepted_by"),
            created_at=req.get("created_at"),
            updated_at=req.get("updated_at"),
            teacher_name=teacher_name
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create request: {str(e)}"
        )


@router.put("/{request_id}/accept", response_model=SubstituteRequestResponse)
async def accept_request(request_id: int, accept_data: AcceptRequest):
    """
    Accept a pending substitute request.
    """
    supabase = get_supabase()
    
    try:
        # Check if request exists and is pending
        check_result = supabase.table("substitute_requests")\
            .select("*")\
            .eq("id", request_id)\
            .execute()
        
        if not check_result.data or len(check_result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        if check_result.data[0]["status"] != "pending":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Request is not available for acceptance"
            )
        
        # Verify acceptor exists and get their name
        acceptor_result = supabase.table("users")\
            .select("id, name")\
            .eq("id", accept_data.teacher_id)\
            .execute()
        
        if not acceptor_result.data or len(acceptor_result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Accepting teacher not found"
            )
        
        acceptor_name = acceptor_result.data[0]["name"]
        original_request = check_result.data[0]
        
        # Accept the request
        result = supabase.table("substitute_requests")\
            .update({
                "status": "accepted",
                "accepted_by": accept_data.teacher_id,
                "updated_at": "now()"
            })\
            .eq("id", request_id)\
            .execute()
        
        if not result.data or len(result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to accept request"
            )
        
        req = result.data[0]
        
        # Notify the original requester that their request was accepted
        await notify_user(
            user_id=original_request["teacher_id"],
            title="✅ Request Accepted!",
            body=f"{acceptor_name} will cover your {req['subject']} class on {req['date']} at {req['time']}",
            data={
                "type": "request_accepted",
                "request_id": request_id,
            }
        )
        
        return SubstituteRequestResponse(
            id=req["id"],
            teacher_id=req["teacher_id"],
            subject=req["subject"],
            date=req["date"],
            time=req["time"],
            duration=req["duration"],
            classroom=req["classroom"],
            notes=req.get("notes"),
            status=req["status"],
            accepted_by=req.get("accepted_by"),
            created_at=req.get("created_at"),
            updated_at=req.get("updated_at"),
            acceptor_name=acceptor_name
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to accept request: {str(e)}"
        )


@router.put("/{request_id}/cancel", response_model=SubstituteRequestResponse)
async def cancel_request(request_id: int, cancel_data: CancelRequest):
    """
    Cancel a substitute request (only by the teacher who created it).
    """
    supabase = get_supabase()
    
    try:
        # Check if request exists and belongs to the teacher
        check_result = supabase.table("substitute_requests")\
            .select("*")\
            .eq("id", request_id)\
            .eq("teacher_id", cancel_data.teacher_id)\
            .execute()
        
        if not check_result.data or len(check_result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found or unauthorized"
            )
        
        original_request = check_result.data[0]
        
        # Cancel the request
        result = supabase.table("substitute_requests")\
            .update({
                "status": "cancelled",
                "updated_at": "now()"
            })\
            .eq("id", request_id)\
            .execute()
        
        if not result.data or len(result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to cancel request"
            )
        
        req = result.data[0]
        
        # If request was accepted by someone, notify them about cancellation
        if original_request.get("accepted_by"):
            await notify_user(
                user_id=original_request["accepted_by"],
                title="❌ Request Cancelled",
                body=f"The substitute request for {req['subject']} on {req['date']} has been cancelled",
                data={
                    "type": "request_cancelled",
                    "request_id": request_id,
                }
            )
        
        return SubstituteRequestResponse(
            id=req["id"],
            teacher_id=req["teacher_id"],
            subject=req["subject"],
            date=req["date"],
            time=req["time"],
            duration=req["duration"],
            classroom=req["classroom"],
            notes=req.get("notes"),
            status=req["status"],
            accepted_by=req.get("accepted_by"),
            created_at=req.get("created_at"),
            updated_at=req.get("updated_at")
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to cancel request: {str(e)}"
        )


@router.delete("/{request_id}")
async def delete_request(request_id: int, teacher_id: int):
    """
    Delete a substitute request (only by the teacher who created it).
    Pass teacher_id as query parameter: /api/requests/{request_id}?teacher_id=1
    """
    supabase = get_supabase()
    
    try:
        # Check if request exists and belongs to the teacher
        check_result = supabase.table("substitute_requests")\
            .select("*")\
            .eq("id", request_id)\
            .eq("teacher_id", teacher_id)\
            .execute()
        
        if not check_result.data or len(check_result.data) == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found or unauthorized"
            )
        
        # Delete the request
        supabase.table("substitute_requests").delete().eq("id", request_id).execute()
        
        return {"message": "Request deleted successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete request: {str(e)}"
        )
